common/djangoapps/student/models.py

```diff
@@ -1030,32 +1032,37 @@ def get_or_create_enrollment(cls, user, course_key):
         if user.id is None:
             user.save()
 
-        try:
-            enrollment, created = cls.objects.get_or_create(
-                user=user,
-                course_id=course_key,
-            )
+        # Even with MySQL set to use the READ COMMITTED isolation level, get_or_create sometimes
+        # raises an IntegrityError and the object in question doesn't appear in a subsequent get()
+        # call. Paired with non_atomic_requests, the retry logic here is meant to prevent IntegrityErrors
+        # from causing authenticated requests to fail.
+        retry = 0
+        while True:
+            # Exponential backoff. Keep sleep time short to prevent running out of available workers.
+            countdown = 2 ** retry * COEFFICIENT
 
-            # If we *did* just create a new enrollment, set some defaults
-            if created:
-                enrollment.mode = CourseMode.DEFAULT_MODE_SLUG
-                enrollment.is_active = False
-                enrollment.save()
-
-        except IntegrityError:
-            log.info(
-                (
-                    "An integrity error occurred while getting-or-creating the enrollment"
-                    "for course key %s and student %s. This can occur if two processes try to get-or-create "
-                    "the enrollment at the same time and the database is set to REPEATABLE READ. We will try "
-                    "committing the transaction and retrying."
-                ),
-                course_key, user
-            )
-            enrollment = cls.objects.get(
-                user=user,
-                course_id=course_key,
-            )
+            try:
+                log.info('Attempting to get or create enrollment for user [%s]. This is retry [%d].', user.username,
+                         retry)
+                enrollment, __ = cls.objects.get_or_create(
+                    user=user,
+                    course_id=course_key,
+                    defaults={
+                        'mode': CourseMode.DEFAULT_MODE_SLUG,
+                        'is_active': False
+                    }
+                )
+
+            except IntegrityError:
+                retry += 1
+                if retry <= MAX_RETRIES:
```

Comment:
@schenedx  We need this if condition here to determine whether we are going to raise error or make retry. You can find similar fix [here](https://github.com/edx/programs/pull/138/files#diff-be9c6bebddcba254a162a11f21ada6cbR71).